R/guides-.R

```diff
@@ -458,49 +453,96 @@ Guides <- ggproto(
     invisible()
   },
 
-  # Loop over every guide, let them draw their grobs
-  draw = function(self, theme, position, direction) {
-    Map(
-      function(guide, params) guide$draw(theme, position, direction, params),
-      guide  = self$guides,
-      params = self$params
-    )
-  },
-
+  # The `Guides$assemble()` method is called in ggplot_gtable (plot-build.R) and
+  # applies the styling from the theme to render each guide and package them
+  # into guide boxes.
+  #
+  # The procedure is as follows
+  #
+  # 1. Guides$draw()
+  #      for every guide object, draw one grob,
+  #      then group the grobs in a list per position
+  #
+  # 2. Guides$package_box()
+  #      for every position, collect all individual guides and arrange them
+  #      into a guide box which will be inserted into the main gtable
   # Combining multiple guides in a guide box
-  assemble = function(self, theme, position) {
+  assemble = function(self, theme) {
 
     if (length(self$guides) < 1) {
       return(zeroGrob())
     }
 
-    position  <- legend_position(position)
-    if (position == "none") {
+    default_position <- theme$legend.position %||% "right"
+    if (length(default_position) == 2) {
+      default_position <- "inside"
+    }
+    if (default_position == "none") {
       return(zeroGrob())
     }
-    default_direction <- if (position == "inside") "vertical" else position
 
     theme$legend.key.width  <- theme$legend.key.width  %||% theme$legend.key.size
     theme$legend.key.height <- theme$legend.key.height %||% theme$legend.key.size
-    theme$legend.box       <- theme$legend.box       %||% default_direction
-    theme$legend.direction <- theme$legend.direction %||% default_direction
-    theme$legend.box.just  <- theme$legend.box.just  %||% switch(
-      position,
-      inside     = c("center", "center"),
-      vertical   = c("left",   "top"),
-      horizontal = c("center", "top")
-    )
 
-    grobs <- self$draw(theme, position, default_direction)
-    if (length(grobs) < 1) {
-      return(zeroGrob())
-    }
+    grobs <- self$draw(theme, default_position)
 
     # Set spacing
     theme$legend.spacing   <- theme$legend.spacing    %||% unit(0.5, "lines")
     theme$legend.spacing.y <- theme$legend.spacing.y  %||% theme$legend.spacing
     theme$legend.spacing.x <- theme$legend.spacing.x  %||% theme$legend.spacing
 
+    Map(
+      grobs    = grobs,
+      position = names(grobs),
+      self$package_box,
+      MoreArgs = list(theme = theme)
+    )
+  },
+
+  # Render the guides into grobs
+  draw = function(self, theme, default_position = "right",
+                  params = self$params, guides = self$guides) {
+    positions <- vapply(
+      params, function(p) p$position[1] %||% default_position, character(1)
+    )
```

Comment:
Is this really necessary since the `Guide$draw()` method already takes it's own position over the one supplied

Reply:
This isn't for `Guide$draw()` method, but for the `split(grobs, position)` line after drawing the grobs.